features/cryptocell/FEATURE_CRYPTOCELL310/aes_alt.c

```diff
@@ -0,0 +1,329 @@
+                    const unsigned char input[16],
+                    unsigned char output[16] )
+{
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    if( ( mode == MBEDTLS_AES_ENCRYPT  && ctx->CC_cipherFlag !=  SASI_AES_ENCRYPT ) ||
+        ( mode == MBEDTLS_AES_DECRYPT && ctx->CC_cipherFlag !=  SASI_AES_DECRYPT ) )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    return( CC_aes_cipher( ctx, mode, SASI_AES_MODE_ECB, 16, NULL, 0, input, output ) );
+
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                           int mode,
+                           size_t length,
+                           unsigned char iv[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    if( ctx == NULL )
+        return ( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+
+    if( length % SASI_AES_BLOCK_SIZE_IN_BYTES )
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+
+    if( ( mode != MBEDTLS_AES_ENCRYPT  || ctx->CC_cipherFlag !=  SASI_AES_ENCRYPT ) &&
+        ( mode != MBEDTLS_AES_DECRYPT || ctx->CC_cipherFlag !=  SASI_AES_DECRYPT ) )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    return( CC_aes_cipher( ctx, mode, SASI_AES_MODE_CBC, length, iv, 16, input, output ) );
+}
+
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_CTR)
+int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+                           size_t length,
+                           size_t *nc_off,
+                           unsigned char nonce_counter[16],
+                           unsigned char stream_block[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    int ret = 0;
+    int n = *nc_off, c, i;
+    size_t j;
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    if ( *nc_off )
+    {
+        /* handle corner case where we are resuming a previous encryption,
+         * and we are resuming within current cipher stream(stream_block) */
+        while( n != 0 )
+        {
+            c = *input++;
+            *output++ = (unsigned char)( c ^ stream_block[n] );
+            n = ( n + 1) & 0x0F;
+            if( length > 0)
+                --length;
+        }
+        /*
+         * Increase the nonce_counter by 1 since we now passed one block
+         */
+          for( i = 16; i > 0; i-- )
+              if( ++nonce_counter[i - 1] != 0 )
+                  break;
+    }
+    if( CC_aes_cipher( ctx, MBEDTLS_AES_ENCRYPT, SASI_AES_MODE_CTR,
+                       length, nonce_counter, SASI_AES_IV_SIZE_IN_BYTES, input, output ) != 0 )
+    {
+        ret = -1;
+    }
+    if( ( ( length % SASI_AES_BLOCK_SIZE_IN_BYTES ) != 0 ) && ret == 0 )
+    {
+        /* in case the length is not aligned, generate stream block for resuming
+         * increase nonce_block to the correct value*/
```

Comment:
Could we fail with hardware unsupported or invalid parameter when alignment is not correct?

If not, maybe it's smaller in code size to copy the block to an aligned location, perform the hardware operation, and then copy back.

Reply:
It's bad wording. It's not a matter of memory alignment. It's a matter of length which is a multiple of AES block size (16). This functionality generates the stream block for resuming on next call( if there is), for generating the first block on next block ( since `*nc_off` is not zero in this case)

Reply:
I think that returning unsupported feature in this case will result in a major functionality decrease, on the expense of minor flash set back, IMHO

Reply:
What's the use case exactly? Why this is a major functionality decrease? Don't users usually use AES with a length that's a multiple of the block length?

Reply:
not for CTR.
CTR is streaming mode, which you can send any length of data, and pause the streaming in the middle, resume the streaming from same location, or any other arbitrary location

Reply:
@Patater CTR mode is a stream cipher based on top of a block operation. From the user's perspective, partial inputs can have any byte length (even any bit length in theory). But under the hood, the operation involving the key works on 128 bits at a time (for a cipher with 128-bit blocks). If a call completes with a total input size that isn't a multiple of the block size, something in the stack has to remember the partial state. The hardware can't do this by itself (it would need to store the state of all ongoing calls). Either the hardware stores the partial state in an opaque form that the driver just keeps in memory, or the hardware works on 128 bits at a time and it's up to the driver to compensate. Usually, it's the latter: if the “hardware” is doing it, it means that some software layer deeper down is doing it.

Reply:
@Patater In addition to what was mentioned by @gilles-peskine-arm CC supports any byte length as input, but it doesn't support byte offset.
As you can see from previous call to CC driver([line 284](https://github.com/ARMmbed/mbed-os/pull/10907/files#diff-b425882d2559838653ea2aa1b992727dR284)), the full length ( after reducing `nc_off` ), no matter what the length is. However, if the length is not a multiple of 16, then we need to calculate `stream_block` for next API call.